server/hotrod/src/main/java/org/infinispan/server/hotrod/ClientListenerRegistry.java

```diff
@@ -388,16 +390,38 @@ void sendEvent(byte[] key, byte[] value, long dataVersion, CacheEntryEvent event
          if (isTrace)
             log.tracef("Queue event %s, before queuing event queue size is %d", remoteEvent, eventQueue.size());
 
-         boolean waitingForFlush = !ch.isWritable();
-         try {
-            eventQueue.put(remoteEvent);
-         } catch (InterruptedException e) {
-            throw new CacheException(e);
-         }
+         EventLoop loop = ch.eventLoop();
+         // We shouldn't be in the event loop, but just in case we can't get stuck blocking on putting into the queue
+         // so we offer and try to catch up on events if possible
+         if (loop.inEventLoop()) {
```

Comment:
IMO we could just throw an exception instead of complicating the code

Reply:
Unfortunately we can't do that really. The problem is someone could add a sync listener after the `CacheRequestProcessor` has already checked and not found one which could cause it possibly to fire in the event loop.

Reply:
TBH I'm still a little confused why we would ever keep the event loop busy while we are running any cache operation. I mean that prevents any other thread from reading from the channel, so the server is going to spend a lot of time twiddling its thumbs unless there are lots of clients per server. I understand that we wouldn't want to incur a context switch to process short requests, but couldn't we allow another thread to start reading from/writing to the channel while we are doing some other work?

Reply:
That is just the way @rvansa wrote it when he changed it over to share the connection.

I originally wrote it the way you said, except gets and ping were exceptions allowing reads on the io thread.

https://github.com/infinispan/infinispan/blob/12839273381601249956986f1987d7ad8758e844/server/hotrod/src/main/java/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.java#L47

When things were more using the async stack was when Radim felt it would be better off allowing all operations that could be async to do this. This PR/bug is a side effect of missing a case :)

I am not sure if what we have is better or worse atm than before (assuming no bugs with blocking). But maybe write operations are too big even when CPU only bound?

Reply:
We should trigger a perf ack job with and without passing requests to another thread pool again, but I'd argue that even the serialization of read results could be heavy enough to justify doing it on another thread.